### PR TITLE
[Security] Removes Doctrine user provider assumption that an invalid User can't be handled at all

### PR DESCRIPTION
**Pull request by iisisrael**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | none

Since Symfony 2.0, if the Doctrine Bridge `EntityUserProvider` detects a user without a valid ID, it throws an `\InvalidArgumentException`, which is not caught by the Firewall `ContextListener` when iterating over all available user providers.  This assumes that an environment cannot be configured to handle such a case with a secondary firewall authenticator.

By throwing and subsequently catching a Symfony namespaced exception instead, the firewall handling can be customized with an additional user provider that throws a `UsernameNotFoundException` instead of an `UnsupportedUserException`.  The authentication then fails gracefully, allowing a secondary authenticator to make an attempt.

The real world scenario for this is where we have an API firewall that handles direct REST requests and AJAX requests from authenticated sessions (using the jQuery `$.ajax()` option `xhrFields: { withCredentials: true }`).  Without this change, the first AJAX request after authenticating in a browser fails on the `\InvalidArgumentException`.  With this change, our custom API user provider can throw a `UsernameNotFoundException` (since it is allowed to try), and our `simple_preauth` authenticator is then allowed to find the user (identifying data stored in a cookie).

I'm not sure if there's even an argument for calling this double authentication approach "unusual" as opposed to "unorthodox".  Our alternative is to refactor our front end to make stateless REST requests or perhaps add a controller layer with a separate route / firewall scheme that passes the logic handling to the API controllers.  I thought it worth submitting this at least as a conversation starter.

### CODE REVIEW

File: CHANGELOG-2.7.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ in 2.7 minor versions.
 To get the diff for a specific change, go to https://github.com/symfony/symfony/commit/XXX where XXX is the change hash
 To get the diff between two versions, go to https://github.com/symfony/symfony/compare/v2.7.0...v2.7.1
 
+* 2.7.41
+ * [Security] Removes Doctrine user provider assumption that an invalid User can't be handled at all (iisisrael)
```
Comment (nicolas-grekas) [inline CHANGELOG-2.7.md]: changelogs are automatically generated, no need to change this (please revert)

File: CONTRIBUTORS.md [hunk 2]
```diff
@@ -1841,3 +1841,4 @@ Symfony is the result of the work of many people who made the code better
  - Matej Žilák (teo_sk)
  - Vladislav Vlastovskiy (vlastv)
  - RENAUDIN Xavier (xorrox)
+ - Israel Carberry (iisisrael)
```
Comment (nicolas-grekas) [inline CONTRIBUTORS.md]: same: autogenerated

**Review by nicolas-grekas — COMMENTED** [review]
shouldn't this qualify as a new feature?

**Comment by xabbuh** [comment]
I wonder why we catch all these exceptions and do not check first instead if the provider actually supports the given user class.

**Comment by iisisrael** [comment]
@nicolas-grekas I wasn't sure if this should be considered a new feature or bug fix, but the more I think about it, I think it's a bug fix.  The `EntityUserProvider` really should play nicely, and not assume it can halt the entire firewall authentication before other providers have a say.  I'm open to resubmitting as a new feature, anyway.

**Comment by chalasr** [comment]
I'm not comfortable with suddenly catching any security InvalidArgumentException from any provider.

@xabbuh I think you're right about calling `supportsClass()` in first place, but it would only allow to stop catching `UnsupportedUserException` right?

**Comment by xabbuh** [comment]
@chalasr Yes, but in this case the provider claimed to support the class. So we needed to call the `refreshUser()` method anyway.

**Comment by iisisrael** [comment]
@chalasr I think changing the exception thrown from a root `\InvalidArgumentException` to the Symfony namespaced one to address any BC concerns, which is what I think you're thinking of:
* If anyone is already catching the `\InvalidArgumentException`, they will still catch the namespaced one.
* If anyone has a custom entity user provider that is throwing an `\InvalidArgumentException` for some other reason, the `ContextListener` will still not catch it.

@xabbuh That would make sense, but not so much a bug fix. The intent here is to fix the user provider, and I considered throwing an `UnsupportedUserException` for the non-persisted state as well, but this seemed more semantically accurate.  Hence the change to the `ContextListener` as well.

**Comment by wouterj** [comment]
I agree with @chalasr that it's a bit scary to suddently change the behaviour of throwing this exception. I think it would be better to instead throw a `UnsupportedUserException` in the doctrine user provider (as I agree with you as well, that the current behavior is unexpected and different from the core user providers).

However, this can be considered a BC break. It can be fixed by creating a new exception class (`UnsupportedUserStateException` whatever) that extends `\InvalidArgumentException` and implements the security exception. However, I think that's too much hassle for a very minor BC break.

**Comment by fabpot** [comment]
Moving the milestone to next for discussion. For the record, I'm 👎 to add a catch to the generic `InvalidArgumentException` class as this is a BC break.

**Comment by stof** [comment]
Can you explain us in which case you considered it valid to refresh a user entity which was not containing the id ?

Making it an error case was done on purpose to make the developer aware of their serialization mistake.

**Comment by iisisrael** [comment]
@stof , an AJAX request is being made from the primary domain to the "api" subdomain, which also allows for persistent sessions, not strictly REST / OAuth requests.  After logging in on the primary domain, the user still hasn't started a session on the api subdomain, so authentication is transported in a cookie available to both domains.

On the first AJAX request, the first authenticator will fail, but our second authenticator in the chain is aware that the cookie may exist.  So even if a user session is not active and a user ID is not detected by the Doctrine Bridge `EntityUserProvider`, it should still be possible to pass responsibility down the chain of authenticators.  This would allow our custom authenticator, which knows to check the cookie in that case, to do its thing.

After the first AJAX request, and if authenticated by the cookie, the user session has started and this is no longer an issue.  (We also handle the flip side if a user logs out on the primary domain and then accesses the API via AJAX, or logs in as a different user in the interim, but that's unrelated to this issue.)

**Comment by stof** [comment]
@iisisrael you are talking about authenticators here, but this PR is about the user provider layer, not about the authenticator system. I think that your own use case might already work without any change in Symfony's DoctrineUserProvider depending on how you implement it.

Throwing an InvalidArgumentException in case a serialized user being reloaded does not include the primary key was done on purpose as part of the DX initiative as this indicates a bad configuration in all cases we saw (generally a bad configuration of the serialization of the user).

So before doing this change, I suggest investigating your use case more to make an informed decision (this PR is known to decrease DX for configuration mistakes in standard use cases)

**Comment by chalasr** [comment]
@iisisrael Any news on this? 
Should we close?

**Comment by iisisrael** [comment]
@chalasr Sure, I'll close it.

**Pull request closed without merging**